### ISSUE TITLE
`@JsonProperty` on constructor parameter changes default field serialization order

### ISSUE THREAD

**Issue by pjfanning** [FasterXML/jackson-databind#3928]
### Discussed in https://github.com/FasterXML/jackson-databind/discussions/3925

<div type='discussions-op-text'>

<sup>Originally posted by **safdevke** May 12, 2023</sup>
Hi, this is a trivial case that I came across. Using nested records changes the order of the properties in the serialized output. Here's a simple example:

```java
public record NestedRecordOne(
    String id,

    String email,

    @JsonProperty(value = "yikes!")
    NestedRecordTwo nestedRecordTwo) {}
```

```java
public record NestedRecordTwo (
    String id,

    String passport) {}
```

```java
public static void main(String[] args) throws IOException {

        ObjectMapper objectMapper = new ObjectMapper();
        NestedRecordTwo nestedRecordTwo = new NestedRecordTwo("2", "111110");
        NestedRecordOne nestedRecordOne = new NestedRecordOne("1", "test@records.com", nestedRecordTwo);

        objectMapper.writeValue(new File("nested-record.json"), nestedRecordOne);
    }
```

The output is:
```json
{
    "yikes!": {
        "id": "2",
        "passport": "111110"
    },
    "id": "1",
    "email": "test@records.com"
}
```

Performing the same test on classes seems to work fine:
```java
public class NestedClassOne {

    private String id;

    private String name;

    @JsonProperty(value = "all is fine!")
    private NestedClassTwo nestedClassTwo;

    // All args constructor + getters + setters
}
```

```java
public class NestedClassTwo {

    private String id;

    // All args constructor + getters + setters
}
```
```java
    NestedClassOne nestedClassOne = new NestedClassOne("1", new NestedClassTwo("2"), "test class");
    
    objectMapper.writeValue(new File("nested-class.json"), nestedClassOne);
```
Produces:
```json
{
    "id": "1",
    "name": "test class",
    "all is fine!": {
        "id": "2"
    }
}
```
This could be easily fixed by using ```@JsonPropertyOrder```. I wonder why it behaves this way for records. I didn't think this warranted creating an issue.</div>

**Comment by pjfanning** [comment]
@cowtowncoder I tested this and created a PR to add a test for the case where no annotations are used (and the order is correct - the field order in the record). When the JsonProperty is added to one field, then that appears to break the order.

**Comment by yihtserns** [comment]
More accurately, "`@JsonProperty` on constructor parameter interferes with default property order", see https://github.com/FasterXML/jackson-databind/discussions/3925#discussioncomment-5893205.

**Comment by pjfanning** [comment]
fix will be in 2.16.0
